File: examples/rust/examples/client_schema_management.rs
//! Schema Management Example
//!
//! This example demonstrates basic schema management operations.
//!
//! Prerequisites:
//! - Run the ekoDB server: `make run`
//!
//! Run with: `cargo run --example client_schema_management`

use ekodb_client::{Client, FieldTypeSchema, Record, Schema};
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Load environment variables
    dotenv::dotenv().ok();

    println!("=== ekoDB Schema Management Example ===\n");

    // Create client
    let client = Client::builder()
        .base_url(
            std::env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string()),
        )
        .api_key(std::env::var("API_BASE_KEY")?)
        .build()?;

    let collection = "schema_client_rust";

    // Step 1: Create a collection with a simple schema
    println!("=== Creating Collection with Schema ===");

    // Delete collection if it exists (cleanup from previous runs)
    let _ = client.delete_collection(collection).await;

    let schema = Schema::new()
        .add_field("title", FieldTypeSchema::new("String").required())
        .add_field("email", FieldTypeSchema::new("String").required())
        .add_field("age", FieldTypeSchema::new("Integer"))
        .add_field("status", FieldTypeSchema::new("String"));

    client.create_collection(collection, schema).await?;
    println!("✓ Created collection '{}' with schema\n", collection);

    // Step 2: Insert documents that conform to schema
    println!("=== Inserting Valid Documents ===");

    let mut user1 = Record::new();
    user1.insert("title", "Software Engineer");
    user1.insert("email", "alice@example.com");
    user1.insert("age", 28);
    user1.insert("status", "active");

    let inserted = client.insert(collection, user1).await?;
    println!("✓ Inserted user 1: {:?}", inserted.get("id"));

    let mut user2 = Record::new();
    user2.insert("title", "Product Manager");
    user2.insert("email", "bob@example.com");
    user2.insert("age", 32);
    user2.insert("status", "pending");

    let inserted = client.insert(collection, user2).await?;
    println!("✓ Inserted user 2: {:?}\n", inserted.get("id"));

    // Step 3: Get schema
    println!("=== Getting Schema ===");

    let schema = client.get_schema(collection).await?;
    println!("✓ Schema for {}:", collection);
    for (name, field) in &schema.fields {
        println!("  - {}: {}", name, field.field_type);
        if field.required {
            println!("    (required)");
        }
    }
    println!();

    // Step 4: List all collections
    println!("=== Listing Collections ===");

    let collections = client.list_collections().await?;
    println!("✓ Total collections: {}", collections.len());
    println!(
        "  Sample: {:?}",
        collections.iter().take(5).collect::<Vec<_>>()
    );
    println!();

    // Cleanup: Delete the collection
    println!("=== Cleanup ===");
    client.delete_collection(collection).await?;
    println!("✓ Deleted collection\n");

    println!("✓ All schema management operations completed successfully");

    Ok(())
}

File: examples/kotlin/examples/ClientSchemaManagement.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.types.Record
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Schema management example - Advanced schema operations
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_schema_example"
    
    println("=== ekoDB Kotlin Client - Schema Management Example ===\n")
    
    try {
        // Step 1: Create collection with schema
        println("=== Creating Collection with Schema ===")
        
        // Delete if exists
        try {
            client.deleteCollection(collection)
        } catch (e: Exception) {
            // Ignore if doesn't exist
        }
        
        val schema = buildJsonObject {
            put("fields", buildJsonObject {
                put("title", buildJsonObject {
                    put("field_type", "String")
                    put("required", true)
                })
                put("email", buildJsonObject {
                    put("field_type", "String")
                    put("required", true)
                })
                put("age", buildJsonObject {
                    put("field_type", "Integer")
                    put("required", false)
                })
                put("status", buildJsonObject {
                    put("field_type", "String")
                    put("required", false)
                })
            })
        }
        
        client.createCollection(collection, schema)
        println("✓ Created collection '$collection' with schema\n")
        
        // Step 2: Insert valid documents
        println("=== Inserting Valid Documents ===")
        
        val user1 = Record.new()
            .insert("title", "Software Engineer")
            .insert("email", "alice@example.com")
            .insert("age", 28)
            .insert("status", "active")
        
        val inserted1 = client.insert(collection, user1)
        println("✓ Inserted user 1: ${inserted1["id"]}")
        
        val user2 = Record.new()
            .insert("title", "Product Manager")
            .insert("email", "bob@example.com")
            .insert("age", 32)
            .insert("status", "pending")
        
        val inserted2 = client.insert(collection, user2)
        println("✓ Inserted user 2: ${inserted2["id"]}\n")
        
        // Step 3: Get schema
        println("=== Getting Schema ===")
        
        val retrievedSchema = client.getSchema(collection)
        println("✓ Schema for $collection:")
        val fields = retrievedSchema["fields"]
        println("  Fields: $fields\n")
        
        // Step 4: List all collections
        println("=== Listing Collections ===")
        
        val collections = client.listCollections()
        println("✓ Total collections: ${collections.size}")
        println("  Sample: ${collections.take(5)}\n")
        
    } finally {
        // Cleanup
        println("=== Cleanup ===")
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n✓ All schema management operations completed successfully")
    }
}
